datafusion/optimizer/src/unwrap_cast_in_comparison.rs

```diff
@@ -766,6 +818,105 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_try_cast_to_type_timestamps() {
+        for time_unit in [
+            TimeUnit::Second,
+            TimeUnit::Millisecond,
+            TimeUnit::Microsecond,
+            TimeUnit::Nanosecond,
+        ] {
+            let utc = Some("+0:00".to_string());
+            // No timezone, utc timezone
+            let (lit_tz_none, lit_tz_utc) = match time_unit {
```

Comment:
A note on the timezone handling is that `expect_cast` added in https://github.com/apache/arrow-datafusion/pull/4147 and used in this PR explicitly verifies that  calling the arrow `cast` kernel produces the same result as the code in this PR.